Efficient Implementation of Self

rachitnigam opened the pull request.

rachitnigam committed:
fix

rachitnigam committed:
words on VLIW machines and bril VLIW machine

rachitnigam committed:
outline

rachitnigam committed:
schedule

rachitnigam committed:
more words

sgpthomas committed:
Wrote list scheduling section

sgpthomas committed:
Merge branch 'blog-2' of github.com:rachitnigam/cs6120 into blog-2

rachitnigam committed:
conclusion

sgpthomas committed:
small cleanup

sgpthomas committed:
Merge branch 'blog-2' of github.com:rachitnigam/cs6120 into blog-2

rachitnigam committed:
reword list scheduling

sgpthomas committed:
superblock scheduling + example fixes

sgpthomas committed:
Merge branch 'blog-2' of github.com:rachitnigam/cs6120 into blog-2

rachitnigam committed:
self blog

rachitnigam committed:
finish blog

sampsyo commented on content/blog/2019-10-28-self/index.md:
```diff
@@ -0,0 +1,176 @@
+
+[An Efficient Implementation of Self, a Dynamically-Typed Object-Oriented
+Language Based on Prototypes][paper] presents techniques for runtime
+compilation, now more commonly referred to as Just In Time (JIT) compilation,
```
```suggestion
compilation, now more commonly referred to as just-in-time (JIT) compilation,
```

It's usually not capitalized, but it usually does have hyphens.

sampsyo commented on content/blog/2019-10-28-self/index.md:
```diff
@@ -0,0 +1,176 @@
+
+The term "dynamic language" is mostly commonly associated with modern scripting
+languages like Python and JavaScript. Self, a much older language developed
+at the famed Xerox PARC labs takes the philosophy of dynamism to a logical
```
```suggestion
at the famed Xerox PARC labs, takes the philosophy of dynamism to a logical
```

Missing a closer for the parenthetical comma.

sampsyo commented on content/blog/2019-10-28-self/index.md:
```diff
@@ -0,0 +1,176 @@
+at the famed Xerox PARC labs takes the philosophy of dynamism to a logical
+extreme -- *everything* in self is a message to an object. This includes
+Java-like method calls on object *as well as* control structures like loops
+and conditionals.
```
It might be worth mentioning that JavaScript is directly inspired by Self (and Scheme), which is why there will be some suspicious similarities here.

sampsyo commented on content/blog/2019-10-28-self/index.md:
````diff
@@ -0,0 +1,176 @@
+if (x) 1 else 2
+```
+
+in Self parlance (written with JavaScript-esque syntax) looks like this:
````
The above is a little hard to understand for two reasons:

- It's not clear which of these is supposed to be JavaScript-like. I think it's the second one. But then it's not 100% sure what "language" the first snippet is written in—I think it's just a notional, abstract language, but maybe it's worth mentioning that it's not supposed to be Self or Self-esque JS.
- In JS, at least `if (x) 1 else 2` isn't terribly meaningful because `if` isn't an expression, unlike `x ? 1 : 2`. Maybe that's fine, though—this does make sense in, like, Scala or Rust.

sampsyo commented on content/blog/2019-10-28-self/index.md:
````diff
@@ -0,0 +1,176 @@
+x.If({ true: () => 1, false: () => 2 })
+```
+
+The runtime behavior of this program is invoking the method `IfTrue` on the
````
```suggestion
The run-time behavior of this program is invoking the method `IfTrue` on the
```
https://capra.cs.cornell.edu/styleguide/#runtime

sampsyo commented on content/blog/2019-10-28-self/index.md:
```diff
@@ -0,0 +1,176 @@
+and execute the "thunk" (a function with no argument) corresponding to the true
+or the false branch.  Note that `x` is not required to be `true` or `false`.
+Any Self object can define the `If` method and specify its conditional
+behavior.
```
This paragraph switches a few times between several names for the same method: `If`, `IfTrue`, and `IfTrue:IfFalse:`. Maybe pick one?

sampsyo commented on content/blog/2019-10-28-self/index.md:
```diff
@@ -0,0 +1,176 @@
+
+The ideas described in the paper have influenced the design of modern JITs for
+JavaScript which use a lot of [similar features][jit] to improve JavaScript execution
+times.
```
Not just JavaScript compilers—basically every JIT in use today descends from this set of ideas. (Even ones for more static-seeming languages like Java and C#.)

sampsyo commented on content/blog/2019-10-28-self/index.md:
```diff
@@ -0,0 +1,176 @@
+A naive layout scheme for objects would copy all fields from a prototype and
+end up wasting a lot of space describing potentially shared behaviors. The Self
+compiler minimizes space usage of _clones_ derived from the same prototype by
+using _clone families_. A cloned object only stores its modifiable in its
```
Missing noun after "modifiable."

sampsyo commented on content/blog/2019-10-28-self/index.md:
```diff
@@ -0,0 +1,176 @@
+end up wasting a lot of space describing potentially shared behaviors. The Self
+compiler minimizes space usage of _clones_ derived from the same prototype by
+using _clone families_. A cloned object only stores its modifiable in its
+object and points to the clone family for its prototype. If the instance ever
```
I'm not 100% sure what "in its object" means—maybe introduce something like an "allocation" to mean the actual memory for laying out fields, as opposed to "object" which is more abstract?

sampsyo commented on content/blog/2019-10-28-self/index.md:
```diff
@@ -0,0 +1,176 @@
+semantics and propagate behavior changes to all clones of a prototype.
+
+<img src="without-maps.png" alt="drawing" width="300"/>
+<img src="with-maps.png" alt="drawing" width="300"/>
```
These figures are a little context free—maybe include a sentence or two describing what we're looking for in these pictures?

sampsyo commented on content/blog/2019-10-28-self/index.md:
```diff
@@ -0,0 +1,176 @@
+
+The Self compiler performs two optimizations, message _inlining_ and
+_splitting_ to enable efficient execution. Message _inlining_ works in a similar
+fashion to customized compilation -- If the type of a receiver is know, either
```
```suggestion
fashion to customized compilation -- if the type of a receiver is known, either
```

sampsyo commented on content/blog/2019-10-28-self/index.md:
```diff
@@ -0,0 +1,176 @@
+method body at the call location. When type information for an object is lost
+due to control flow splits, _message splitting_ generates specialized code for
+possible receiver types and guards them using type tests. If the type of
+receiver matches, fast code can be executed.
```
To drive this point home, maybe it would be useful to include some example code to illustrate what the program looks like after these optimizations are applied. For example, you could show several imaginary assembly- or C-like listings depicting these versions of a program `x < y`:

- The obvious version, where the code looks up the string `lessThan` in the method dictionary for the object `x` and jumps to the corresponding code with argument `y`.
- The "customized compilation" version, with a type guard and a direct `call` to the `integer.lessThan` function with `y` as the argument.
- The "inlined" version, which still has the type guard but just uses a `lt` machine instruction or whatever to do the comparison—no call necessary.

For bonus points, you could even show imaginary code for the "split" version.

sampsyo commented on content/blog/2019-10-28-self/index.md:
```diff
@@ -0,0 +1,176 @@
+
+### Programming Environment Support
+
+The Self compiler support _incremental recompilation_ and _source-level
```
```suggestion
The Self compiler supports _incremental recompilation_ and _source-level
```

sampsyo commented on content/blog/2019-10-28-self/index.md:
```diff
@@ -0,0 +1,176 @@
+inlined from a clone's parent, updating the parent pointer or the method body
+in the parent should invalidate the specialized code. Since the compiler
+selectively invalidates code, methods that weren't modified can still execute
+with specialized code.
```
This is the weirdest part of Self, and the one that has not stuck with us today: the interactive, live-editable code environment. It's maybe worth pointing out that this is the most anachronistic part of the paper as a result. V8 doesn't have an interactive, persistent, first-class IDE, for example, so it doesn't have to worry about this as much.

sampsyo commented on content/blog/2019-10-28-self/index.md:
```diff
@@ -0,0 +1,176 @@
+than the original C programs.
+
+The authors also describes a new metric to measure the performance of an object
+oriented programming language - _Millions of Messages per Second (MiMS)_ which
```
Elsewhere, you have used ` -- ` to separate thoughts, rather than just a single hyphen. Consider sticking to that or just using commas here.

sampsyo commented on content/blog/2019-10-28-self/index.md:
```diff
@@ -0,0 +1,176 @@
+
+The authors also describes a new metric to measure the performance of an object
+oriented programming language - _Millions of Messages per Second (MiMS)_ which
+corresponds to MIPS. To compute the MiMS of a specific virtual machine, divide
```
How does it "correspond" to MIPS? The paper says it's "analogous" to MIPS, which I think is more accurate.

sampsyo reviewed (changes requested):
Looks pretty good; thanks for putting this together! This PR does contain your Project 2 blog post, in case that's easy to resolve.

Although you don't need to hit every point, I notice that there are a few things from the paper that you didn't cover, some of which are historically pretty important:

- Heap organization to facilitate GC ("segregation").
- Inline word tagging for integer & FP values.
- Bytecode (preceded the JVM).

For at least the second two, which are bog-standard JIT implementation techniques today, it might be worth at least briefly mentioning that this paper covers them—and attempting to put them into a modern context.

rachitnigam committed:
Update content/blog/2019-10-28-self/index.md

Co-Authored-By: Adrian Sampson <adrian@radbox.org>

rachitnigam committed:
Update content/blog/2019-10-28-self/index.md

Co-Authored-By: Adrian Sampson <adrian@radbox.org>

rachitnigam committed:
Update content/blog/2019-10-28-self/index.md

Co-Authored-By: Adrian Sampson <adrian@radbox.org>

rachitnigam committed:
Update content/blog/2019-10-28-self/index.md

Co-Authored-By: Adrian Sampson <adrian@radbox.org>

rachitnigam committed:
better explanation for conditionals

rachitnigam committed:
fields

rachitnigam committed:
more fixes

rachitnigam committed:
inlining example

rachitnigam committed:
motivation for prog env

rachitnigam committed:
javascript like

rachitnigam committed:
words for allocation map

rachitnigam committed:
typos

rachitnigam committed:
bytecode

rachitnigam committed:
segregation and tagging

rachitnigam commented:
Thanks for the comments! I've updated the inline comments and added sections on segregation, tagging, and the bytecode. I initially chose to forgo them because I thought they were not contributions of the paper. It is remarkable how many ideas this paper contains!

I'm going to move this to a new PR to remove the commits from my project.

rachitnigam commented:
Moved to #95.
